File: src/main/java/net/theevilreaper/vulpes/generator/controller/BuildInformationHandler.java
package net.theevilreaper.vulpes.generator.controller;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import jakarta.inject.Inject;
import net.theevilreaper.vulpes.generator.domain.client.GithubReleaseService;
import net.theevilreaper.vulpes.generator.domain.release.GitReleaseDTO;
import org.jetbrains.annotations.NotNull;

@Controller("/build")
public class BuildInformationHandler {

    private final GithubReleaseService releaseService;

    /**
     * Creates a new instance of the {@link BuildInformationHandler}
     *
     * @param releaseService the {@link GithubReleaseService} to use for retrieving
     *                       the latest release information.
     */
    @Inject
    public BuildInformationHandler(GithubReleaseService releaseService) {
        this.releaseService = releaseService;
    }

    /**
     * Returns the latest build information with a specific level of detail.
     *
     * @return the latest build information
     */
    @Operation(
            summary = "Get build information",
            operationId = "getBuildData",
            description = "Returns the latest build information including version and creation date.",
            tags = {"build"}
    )
    @ApiResponse(
            responseCode = "200",
            description = "Build information retrieved successfully.",
            content = @Content(
                    mediaType = MediaType.APPLICATION_JSON,
                    schema = @Schema(implementation = GitReleaseDTO.class)
            )
    )
    @Get(value = "/data", produces = MediaType.APPLICATION_JSON)
    @ExecuteOn(TaskExecutors.BLOCKING)
    public @NotNull HttpResponse<GitReleaseDTO> getBuildInformation() {
        GitReleaseDTO latestVersion = this.releaseService.getLatestVersion();
        return HttpResponse.ok(latestVersion);
    }
}

File: src/main/kotlin/net/theevilreaper/vulpes/generator/controller/BuildInformationHandler.kt
package net.theevilreaper.vulpes.generator.controller

import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import jakarta.inject.Inject
import net.theevilreaper.vulpes.generator.properties.GitlabProperties
import org.gitlab4j.api.Constants
import org.gitlab4j.api.GitLabApi
import org.gitlab4j.api.models.PackageFilter
import org.gitlab4j.api.models.Project

@Controller
class BuildInformationHandler @Inject constructor(
    gitLabConfig: GitlabProperties
) {

    private val gitLabApi: GitLabApi = GitLabApi(gitLabConfig.gitlabUrl, gitLabConfig.accessToken)

    @Get("/build/data", produces = [MediaType.APPLICATION_JSON])
    private fun getBuildInformation(): HttpResponse<BuildInformation> {
        val projectApi = gitLabApi.projectApi
        val projects: List<Project> = projectApi.getProjects("aves")

        if (projects.isEmpty()) {
            return HttpResponse.ok(BuildInformation(mapOf("error" to "Project not found")))
        }

        val avesProject = projects.first()
        val filter = PackageFilter().withOrderBy(Constants.PackageOrderBy.CREATED_AT)
            .withSortOder(Constants.SortOrder.DESC)

        val latestReleases = gitLabApi.packagesApi.getPackages(avesProject.id, filter, 1)

        if (!latestReleases.hasNext()) {
            return HttpResponse.ok(BuildInformation(mapOf("error" to "No releases found")))
        }

        val release = latestReleases.first().first()
        val data: MutableMap<String, String> = mutableMapOf()
        data["version"] = release.version
        data["created"] = release.createdAt.toString()
        return HttpResponse.ok(BuildInformation(data))
    }

    data class BuildInformation(
        val data: Map<String, String>
    )
}
